pkg/audit/manager.go

```diff
@@ -213,40 +212,90 @@ func (am *Manager) auditResources(ctx context.Context) ([]*constraintTypes.Resul
 	var responses []*constraintTypes.Result
 	var errs opa.Errors
 
-	for gv, gvKinds := range clusterAPIResources {
-		for kind := range gvKinds {
-			objList := &unstructured.UnstructuredList{}
-			objList.SetGroupVersionKind(schema.GroupVersionKind{
-				Group:   gv.Group,
-				Version: gv.Version,
-				Kind:    kind + "List",
-			})
+	ck, err := am.getAllConstraintKinds()
+	if err != nil {
+		// if no constraint is found with the constraint apiversion, then return
+		am.log.Info("no constraint is found with apiversion", "constraint apiversion", constraintsGV)
+		return nil, nil
+	}
 
-			err := am.client.List(ctx, objList)
+	constraintsList := make(map[string]bool)
+	constraintList := &unstructured.UnstructuredList{}
+	for _, c := range ck {
+		constraintList.SetGroupVersionKind(c)
+		if err = am.client.List(ctx, constraintList); err != nil {
+			am.log.Error(err, "Unable to list objects for gvk", "group", c.Group, "version", c.Version, "kind", c.Kind)
+			continue
+		}
+		for _, constraint := range constraintList.Items {
+			kinds, found, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
 			if err != nil {
-				am.log.Error(err, "Unable to list objects for gvk", "group", gv.Group, "version", gv.Version, "kind", kind)
+				am.log.Error(err, "Unable to return spec.match.kinds field", "group", c.Group, "version", c.Version, "kind", c.Kind)
 				continue
 			}
-
-			for _, obj := range objList.Items {
-				ns := &corev1.Namespace{}
-				if obj.GetNamespace() != "" {
-					if err := am.client.Get(ctx, types.NamespacedName{Name: obj.GetNamespace()}, ns); err != nil {
-						am.log.Error(err, "Unable to look up object namespace", "group", gv.Group, "version", gv.Version, "kind", kind)
+			if found {
+				for _, k := range kinds {
+					kind, ok := k.(map[string]interface{})
+					if !ok {
+						am.log.Error(errors.New("could not cast kind as map[string]"), "kind", k)
+						continue
+					}
+					kindsKind, _, err := unstructured.NestedSlice(kind, "kinds")
+					if err != nil {
+						am.log.Error(err, "Unable to return kinds.kinds field", "group", c.Group, "version", c.Version, "kind", c.Kind)
 						continue
 					}
+					for _, kk := range kindsKind {
+						// adding constraint match kind to list
+						constraintsList[kk.(string)] = true
+					}
 				}
+			} else {
```

Comment:
What is the likelihood that this logic path is hit? If it gets hit, ironically we are in a worse situation as we have extra requests to the API server to gather constraints and a higher probability of introducing bugs due to added code complexity or failing an audit cycle/reporting inaccurate results due to a flaky API server.